avro-codegen/src/main/java/com/linkedin/avroutil1/codegen/SpecificRecordClassGenerator.java

```diff
@@ -391,46 +392,60 @@ protected JavaFileObject generateSpecificRecord(AvroRecordSchema recordSchema, S
-          .addModifiers(Modifier.PUBLIC)
-          .addCode(CodeBlock
-              .builder()
-              .addStatement("$L.write(this, $T.getEncoder(out))", "WRITER$", SpecificRecordGeneratorUtil.CLASSNAME_SPECIFIC_DATA)
-              .build());
-      readExternalBuilder.addAnnotation(Override.class);
-      writeExternalBuilder.addAnnotation(Override.class);
+    //READER$
+    ParameterizedTypeName datumReaderType = ParameterizedTypeName.get(
+        SpecificRecordGeneratorUtil.CLASSNAME_DATUM_READER,
+        TypeVariableName.get(recordSchema.getSimpleName())
+    );
+    classBuilder.addField(
+        FieldSpec.builder(datumReaderType, "READER$", Modifier.PRIVATE, Modifier.STATIC, Modifier.FINAL)
+            .initializer(
+                CodeBlock.of("com.linkedin.avroutil1.compatibility.AvroCompatibilityHelper.newSpecificDatumReader($L, $L, $L)", "SCHEMA$", "SCHEMA$", "MODEL$")
+            )
+            .build()
+    );
 
-      classBuilder.addMethod(readExternalBuilder.build());
-      classBuilder.addMethod(writeExternalBuilder.build());
-    }
+    //readExternal()
+    MethodSpec.Builder readExternalBuilder = MethodSpec.methodBuilder("readExternal")
+        .addException(IOException.class)
+        .addParameter(java.io.ObjectInput.class, "in")
+        .addModifiers(Modifier.PUBLIC)
+        .addCode(CodeBlock.builder().addStatement("$L.read(this, com.linkedin.avroutil1.compatibility.AvroCompatibilityHelper.newBinaryDecoder(in))", "READER$").build());
+    //no @Overrides annotation since under old avro the parent class is not externalizable
+    //TODO - add this is min avro version is high enough
+    //readExternalBuilder.addAnnotation(Override.class);
+    classBuilder.addMethod(readExternalBuilder.build());
+
+    //WRITER$
+    ParameterizedTypeName datumWriterType = ParameterizedTypeName.get(
+        SpecificRecordGeneratorUtil.CLASSNAME_DATUM_WRITER,
+        TypeVariableName.get(recordSchema.getSimpleName())
+    );
+    classBuilder.addField(
+        FieldSpec.builder(datumWriterType, "WRITER$", Modifier.PRIVATE, Modifier.STATIC, Modifier.FINAL)
+            .initializer(
+                CodeBlock.of("com.linkedin.avroutil1.compatibility.AvroCompatibilityHelper.newSpecificDatumWriter($L, $L)", "SCHEMA$", "MODEL$")
+            )
+            .build());
 
+    //writeExternal()
+    MethodSpec.Builder writeExternalBuilder = MethodSpec
+        .methodBuilder("writeExternal")
+        .addException(IOException.class)
+        .addParameter(java.io.ObjectOutput.class, "out")
+        .addModifiers(Modifier.PUBLIC)
+        .addCode(CodeBlock
+            .builder()
+            .addStatement("$L.write(this, com.linkedin.avroutil1.compatibility.AvroCompatibilityHelper.newBinaryEncoder(out))", "WRITER$")
+            .build());
+    //no @Overrides annotation since under old avro the parent class is not externalizable
+    //TODO - add this is min avro version is high enough
```

Comment:
what do you mean by this TODO? in general, curious why we are punting on this (is it difficult / low value, is it not possible right now, etc

Reply:
the overrides annotation has no value. we strip it out to avoid compilation issues under old avro.
we need to spend a more significant amount of time figuring out what "min target avro version" should actually mean.